cmd/queue.go

```diff
@@ -0,0 +1,449 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/jedib0t/go-pretty/v6/table"
+	"github.com/jedib0t/go-pretty/v6/text"
+	"github.com/spf13/cobra"
+
+	sdk "github.com/ivcap-works/ivcap-cli/pkg"
+	a "github.com/ivcap-works/ivcap-cli/pkg/adapter"
+	api "github.com/ivcap-works/ivcap-core-api/http/queue"
+)
+
+func init() {
+	rootCmd.AddCommand(queueCmd)
+
+	// LIST
+	listCommand()
+
+	// READ
+	readCommand()
+
+	// CREATE
+	createCommand()
+
+	// DELETE
+	deleteCommand()
+
+	// ENQUEUE
+	enqueueCommand()
+
+	// DEQUEUE
+	dequeueCommand()
+}
+
+var queueCmd = &cobra.Command{
+	Use:     "queue",
+	Aliases: []string{"q", "queues"},
+	Short:   "Create and manage queues",
+	Long:    `Queues are used to store messages in a sequential order. You can create, read, update, and delete queues using this command. You can also add and remove messages from queues.`,
+}
+
+func listCommand() {
+	listQueueCmd := &cobra.Command{
+		Use:   "list",
+		Short: "List existing queues",
+		RunE:  runListQueueCmd,
+	}
+
+	queueCmd.AddCommand(listQueueCmd)
+	addListFlags(listQueueCmd)
+}
+
+func readCommand() {
+	readQueueCmd := &cobra.Command{
+		Use:     "get [flags] queue_id",
+		Aliases: []string{"read"},
+		Short:   "Fetch details about a single queue",
+		Long: `Fetch details about a single queue. You must provide the ID of the queue you want to read.
+
+An example of reading a queue:
+
+  ivcap queue get urn:ivcap:queue:714e549b-ebab-5dd8-8ebd-2e4b0af76167`,
+		Args: validateReadCommandArgs,
+		RunE: runReadQueueCmd,
+	}
+
+	queueCmd.AddCommand(readQueueCmd)
+}
+
+func createCommand() {
+	createQueueCmd := &cobra.Command{
+		Use:   "create [flags] name",
+		Short: "Create a new queue",
+		Long: `Create a new queue with the specified name. Optionally, you can also provide a description and a policy for the queue.
+
+An example of creating a queue with a description:
+
+  ivcap queue create --description "This is a test queue" test_queue`,
+		Args: validateCreateQueueArgs,
+		RunE: runCreateQueueCmd,
+	}
+
+	queueCmd.AddCommand(createQueueCmd)
+	createQueueCmd.Flags().StringP("description", "d", "", "Description of the queue")
+	addPolicyFlag(createQueueCmd)
+}
+
+func deleteCommand() {
+	deleteQueueCmd := &cobra.Command{
+		Use:     "delete [flags] queue_id",
+		Aliases: []string{"remove"},
+		Short:   "Delete a queue",
+		Long: `Delete a queue with the specified ID. This action is irreversible. All messages in the queue will be lost.
+An example of deleting a queue:
+
+  ivcap queue delete test_queue`,
+		Args: validateDeleteCommandArgs,
+		RunE: runDeleteQueueCmd,
+	}
+	queueCmd.AddCommand(deleteQueueCmd)
+}
+
+func enqueueCommand() {
+	enqueueCmd := &cobra.Command{
+		Use:   "enqueue [flags] schema queue_id message",
```

Comment:
Can we swap `schema` and `queue_id`? For most commands we usually have the resource id right after the sub command.

I would also ingest the message from a file `-f`. Check commands like `service create` which should have implemented that. One can always get it from the command line via `echo "...message..." | ivcap queue enqueue .. -f -`

Reply:
Just thinking, you may want to look at `aspectAddCmd`.  In this case the entityID (aka queueID) is mandatory, the schema is defined either by an optional `--schema` flag or of the message is of type json and has a top-level `$schema` field.

Not sure if you could pretty much reuse most of the code - may require a little refactoring to make it work for both cases - all assuming that the two are really that similar as it looks to me.

Reply:
I appreciate the suggestion, I will take a look and get back to you with any questions I may have.